File: mahaam-api-ts/src/infra/monitor/traffic.repo.ts
import { Injectable } from '@nestjs/common';
import { Traffic } from './monitor.model';
import { DB } from '../db';

export interface TrafficRepo {
  create(traffic: Traffic): Promise<void>;
}

@Injectable()
export class DefaultTrafficRepo implements TrafficRepo {
  async create(traffic: Traffic): Promise<void> {
    setImmediate(async () => {
      try {
        await DB.sql`INSERT INTO x_traffic (id, health_id, method, path, code, elapsed, headers, request, response, created_at) 
	VALUES (${traffic.id}, ${traffic.healthId}, ${traffic.method}, ${traffic.path}, ${traffic.code || null}, ${traffic.elapsed || null}, 
	${traffic.headers || null}, ${traffic.request || null}, ${traffic.response || null}, current_timestamp)`;
      } catch (error) {
        console.error(error);
      }
    });
  }
}

File: mahaam-api-py/infra/monitor/traffic_repo.py
from typing import Protocol
from concurrent.futures import ThreadPoolExecutor
from infra import db
from infra.monitor.monitor_models import Traffic
from infra.validation import ProtocolEnforcer
from infra.log import Log


class TrafficRepo(Protocol):
    def create(self, traffic: Traffic) -> None: ...


class DefaultTrafficRepo(metaclass=ProtocolEnforcer, protocol=TrafficRepo):

    def create(self, traffic: Traffic) -> None:
        def insert_traffic():
            try:
                traffic_data = {
                    "id": str(traffic.id),
                    "health_id": str(traffic.health_id),
                    "method": traffic.method,
                    "path": traffic.path,
                    "code": traffic.code,
                    "elapsed": traffic.elapsed,
                    "headers": traffic.headers,
                    "request": traffic.request,
                    "response": traffic.response
                }

                sql = """
                    INSERT INTO x_traffic (id, health_id, method, path, code, elapsed, headers, request, response, created_at) 
                    VALUES(:id, :health_id, :method, :path, :code, :elapsed, :headers, :request, :response, current_timestamp)"""

                db.DB.insert(sql, traffic_data)
            except Exception as e:
                Log.error(f"error creating traffic record: {e}")

        ThreadPoolExecutor(max_workers=1).submit(insert_traffic)
